### ISSUE TITLE
JKL Shuttle Tasten

### ISSUE THREAD

**Issue by fernsehmuell** [Ultraschall/ultraschall-portable#34]
JKL Shuttle Funktionalität, wie in den großen Video Schnittsystemen.

**Comment by rstockm** [comment]
Bei der Benennung des Branches bitte demnächst die #Nummer nach vorne setzen. Siehe Wiki...

**Comment by fernsehmuell** [comment]
Habt ihr Wunschgeschwindigkeiten?
Derzeit 1->2>3>5>8x

Wenn das eine Parabel (0.5x^2-1.5x+3) sein soll wären die nächsten in der Reihe: 
12, 17, 23, 30, 38, 47, 57, 68, 80, 93, 107, 122, 138, 155, 173, 192, 212, 233, 255, 278, 302, 327, 353

**Comment by fernsehmuell** [comment]
Hallo zusammen, die Funktion für die Taste "L" ist jetzt in einem testbaren Zustand. Sie bricht leider derzeit den Folllowmode. Wahrscheinlich, weil ich den Editcursor immer nachziehe. @mespotine kann das der Grund sein? Die Frage ist auch noch, ob das das gewünschte Verhalten ist. Ich habe es eingebaut, damit wenn man ganz am Ende ankommt der Cursor auch am Projektende steht.

**Comment by rstockm** [comment]
Diesen edge-case (sic) finde ich persönlich ja total egal - am Ende freut man sich und macht andere Sachen und wundert sich nicht wo der Cursor steht...

**Comment by rstockm** [comment]
Habe das mal auf `L`gelegt - sobald ich aber versuche die Geschwindigkeit zu erhöhen:

<img width="516" alt="image" src="https://user-images.githubusercontent.com/3195116/37492268-04564b42-28a1-11e8-9f6a-ba2580dc8908.png">

**Comment by fernsehmuell** [comment]
Ich glaube terminate war hier das richtige. die Entscheidung wird übrigens als 260 bzw 516 in der Zeile SRC ... anstatt der 4 gespeichert (Reaper-kb.ini) Also BIT 8 bzw. 9

**Comment by rstockm** [comment]
Hm, ja nee :-)
Wir haben bisher unsere Skripte so gebaut, dass seinigen diesen kb-Hack ausgekommen sind - der @mespotine müsste da was am Start haben?

**Comment by mespotine** [comment]
Einfach in meine kb.ini-Doku schaun, da findet Ihr, wo man das eintragen kann :)

@fernsehmuell 
Kannste nochmal kurz beschreiben, wie das JKL-Management funzt, bzw gedacht ist? Also welche Taste was unter welchen BEdingungen macht? Kanns grad nicht testen, will aber trotzdem meinen Senf zu geben ;)

**Comment by mespotine** [comment]
Und dass beim Ändern der Editcursor-Position sich Follow ausschaltet ist gewollt. Gibt ne Workaround-Funktion in den Helper-Functions dafür, um das auszuschalten, sowie eine Action.
Musste mal schaun, wie das DoppelKlick für Play macht...

**Comment by mespotine** [comment]
Beschreibung grad im andern Issue gefunden:

> 
> Ich habe den Branch JOG-SHUTTLE-#18 dazu angelegt. Sind derzeit die aktuellen Shuttle Script aus meinem AvidMediaComposerStyle Repo.
> 
> Benutzung:
> Scripte sind im Branch auf die J K L Tasten gemapped.
> J= bewege Cursor Rückwärts
> K=Pause
> L= Play
> bei mehrmaligem Drücken erhöht sich die Geschwindigkeit um jeweils eine Stufe.
> Da es in Reaper keinen negative Wiedergabegeschwindigkeit gibt, musste ich da mit einem Backgroundscript arbeiten und es gibt auch keinen Ton.


Für vorspulen gibts da nix? Also fürs vorspulen ohne dass man was hört(also zum stillen an die richtige Stelle fahren)? Oder ist das eh nicht notwendig?

**Comment by fernsehmuell** [comment]
Für das schnelle Vorspulen benutze ich einfach den internen Playbackspeed von Reaper. Kann man bestimmt auch stumm machen, aber ich sehe gerade keinen Grund warum das stört.

**Comment by rstockm** [comment]
Ich habe hier mal einen Schwung Kommentare am Code hinterlassen, ohne es getestet haben zu können.
Normalerweise wäre ich da nicht so picky, aber bei Hintergrundskripten sollten wir alle sehr, sehr gründlich drauf schauen damit wir uns da keine Performance und/oder Stabilitätsprobleme einhandeln.
Generell würde ich da auch etwas mehr Dokumentation gut finden.

https://github.com/Ultraschall/ultraschall-portable/commit/b831dc94c47045b724533584faefb75bbb0db28d#commitcomment-28265502

**Comment by mespotine** [comment]
Läuft das Hintergrundscript eigentlich permanent? Oder kann man das "bei Bedarf" startbar machen, so dass es nur läuft, wenn wir es wirklich brauchen?

**Comment by fernsehmuell** [comment]
Nein, das Script läuft nur, wenn es auch tatsächlich "shuttled". Wenn man "K" drückt, dann wird der externe state auf "0" gestetzt, was den Scripts sagt "bitte beende dich"

**Comment by fernsehmuell** [comment]
Ich habe alle Kommentare im Code durchgearbeitet und dort auch weider kommentiert. Die vier Skripte sind jetzt geändert im Branch.

**Comment by rstockm** [comment]
Das sieht jetzt sehr schön aus, werde nachher testen.

**Comment by rstockm** [comment]
Ein doch relevanter Schönheitsfehler: wenn man mit J rückwärts spult, dann wird followmode abgestellt, soweit richtig. Erreicht man die grenze des Viewports, wird dieser seitenweise geblättert: auch gewünschtes Verhalten.
Beim `l` nach vorne ist dem jedoch nicht so: da bleibt der Viewport stehen und auch ein Drücken von `k` zentriert nicht den View. Gewünschtes Verhalten: der Viewport sollte sich so wie bei `j` mit umblättern.

**Comment by fernsehmuell** [comment]
Gucke ich mir an. Da `j` einfach nur die Playrate verändert, ist das vielleicht ein bekannter Toggle den man umstellen kann?

**Comment by rstockm** [comment]
Hm, nochmal nachgesehen: jain. Das hängt mit unsrem follow-mode zusammen: der Scrollt ja immer flüssig. Wenn er abgeschaltet ist, scrollt gar nichts mehr - auch das ist gewünschtes Verhalten. M.E. braucht es für den JKL-Mode jetzt ein mittelding: eben seitenweises Scrolling, was für den follow-mode aber deaktiviert wurde. Frage ist: warum läuft es beim `j` dennoch sauber, aber nicht beim `k` - eventuell muss man da dann doch im backrounds-Skript noch Viewport-foo von @mespotine verbauen.

**Comment by fernsehmuell** [comment]
@rstockm bitte mal diese Version testen. Ich habe das Verhalten der `k` Taste wieder auf STOP gestellt, da sonst durch 2x drücken die Wiedergabe startet, ich habe aber gemerkt, dass man in der Praxis gerne mal öfter draufdrückt, um auch wirklich zu stoppen. Andere Kleinigkeiten habe ich auch noch geändert (siehe commit)

**Comment by rstockm** [comment]
@fernsehmuell wir haben hier leider doch noch ein Problem, und zwar mit den multiplen Instanzen des forward Skripts. 

Folgendes Verhalten lässt sich reproduzieren:
- wenn man - wie vom Skript vorgeschlagen - "new instance" auswählt beim der Dopplungs-Warnung, so funktioniert zwar alles, mit jeder Erhöhung der Geschwindigkeit kommt jedoch ein laufendes background-Skript hinzu. Diese werden durch `k` dann _nicht_ abgeräumt, nur das jeweils letzte. So hat man nach kurzer Zeit dutzende laufende Skripts - uncool.
- wenn man stattdessen `terminate` auswählt, hat man immer nur ein Skript am laufen, und das wird auch sauber geschlossen. Jetzt hat man aber das Problem, dass der Edit-Cursor bei genau jedem zweiten Gescwindigkeits-Boost stehen bleibt, und nicht mit dem Play-Cursor mitwandert - was zum einen inkonsistentes Verhalten ist, zum anderen aber auch dazu führt, dass wenn man bei dem Zustand mit `k` abstoppt die Cursor beide an der falschen Stelle liegen. Uncool.

Da muss eine Lösung her...

**Comment by lootsy** [comment]
Tach Gemeinde, beim Verfolgen dieses Issues bin ich leider überfordert. :-/

Ich würde gerne mal live, 1:1, vis-a-vis, tete-a-tete mit euch besprechen, ob das auch so geht, dass ich das verstehen kann. ;-)

@fernsehmuell Nimm dir mal am Donnerstag und Montag in Westfeld nichts vor. ;-))

**Comment by fernsehmuell** [comment]
So das mit den zig Instanzen ist gefixed. Jedes Script erzeugt sich beim Start jetzt eine ID (Zähler). Wenn es im KV-Store eine ID gibt die größer als die des Skriptes ist, dann wird das alte Script beendet.
Das letzte Script setzt vor dem Beenden die ID wieder auf "". Das läuft hier ganz prima. Beim ersten mal dauerhaft auf "new Instance" stellen.

**Comment by rstockm** [comment]
Ja, sieht gut aus. Einen Schönheitsfehler finde ich noch: wenn man von `l` direkt nach `j` springt, bleibt noch eine forward-Instanz im Hintergrund aktiv. Das sollte vermutlich nicht so sein?

**Comment by mespotine** [comment]
Wenn die alten Script-Instanzen jetzt in der derzeitigen Implementation beendet werden, werden die neuen Instanzen auch zuverlässig wieder gestartet? 
Sprich, wenn Ihr das über die KB.ini-Einstellungen "Terminate Instance" gelöst habt, kanns sein, dass die alte Instanz nur beendet wird bei erneutem Shortcut-auslösen, aber die neue nicht gestartet...
